swarm/chunk/tag.go

```diff
@@ -0,0 +1,194 @@
+
+package chunk
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	errExists       = errors.New("already exists")
+	errNA           = errors.New("not available yet")
+	errUnknownState = errors.New("unknown state")
+	errNoETA        = errors.New("unable to calculate ETA")
+)
+
+// State is the enum type for chunk states
+type State = uint32
+
+const (
+	SPLIT  State = iota // chunk has been processed by filehasher/swarm safe call
+	STORED              // chunk stored locally
+	SEEN                // chunk previously seen
+	SENT                // chunk sent to neighbourhood
+	SYNCED              // proof is received; chunk removed from sync db; chunk is available everywhere
+)
+
+// Tag represents info on the status of new chunks
+type Tag struct {
+	Uid       uint32    // a unique identifier for this tag
+	Name      string    // a name tag for this tag
+	total     uint32    // total chunks belonging to a tag
+	split     uint32    // number of chunks already processed by splitter for hashing
+	seen      uint32    // number of chunks already seen
+	stored    uint32    // number of chunks already stored locally
+	sent      uint32    // number of chunks sent for push syncing
+	synced    uint32    // number of chunks synced with proof
+	startedAt time.Time // tag started to calculate ETA
+}
+
+// New creates a new tag, stores it by the name and returns it
+// it returns an error if the tag with this name already exists
+func NewTag(uid uint32, s string, total uint32) *Tag {
+	t := &Tag{
+		Uid:       uid,
+		Name:      s,
+		startedAt: time.Now(),
+		total:     total,
+	}
+	return t
+}
+
+// Inc increments the count for a state
+func (t *Tag) Inc(state State) {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SEEN:
+		v = &t.seen
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	atomic.AddUint32(v, 1)
+}
+
+// Get returns the count for a state on a tag
+func (t *Tag) Get(state State) int {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SEEN:
+		v = &t.seen
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	return int(atomic.LoadUint32(v))
+}
+
+// GetTotal returns the total count
+func (t *Tag) Total() int {
+	return int(atomic.LoadUint32(&t.total))
+}
+
+// DoneSplit sets total count to SPLIT count
+// is meant to be called when splitter finishes for input streams of unknown size
+func (t *Tag) DoneSplit() int {
+	total := atomic.LoadUint32(&t.split)
+	atomic.StoreUint32(&t.total, total)
+	return int(total)
+}
+
+// Status returns the value of state and the total count
+func (t *Tag) Status(state State) (int, int, error) {
+	count, seen, total := t.Get(state), int(atomic.LoadUint32(&t.seen)), int(atomic.LoadUint32(&t.total))
+	if total == 0 {
+		return count, total, errNA
+	}
+	switch state {
+	case SPLIT:
```

Comment:
still not right. 

- Both stored and split should be relative to total chunk instances;
- Both sent and synced should be relative to total distinct (stored-seen)

switch state {
	case SPLIT, STORED:
            return count, total, nil 
        case SENT, SYNCED:
            	stored := int(atomic.LoadUint32(&t.stored))
		if stored < total {
                      return count, total-seen, errNA
		}
                return count, total-seen, nil
           }
	}

Reply:
stored will always be smaller than total because we only increment stored on `has, _ = s.Put(); has==true`. or that has to change too?

Reply:
cardinality of `total` right now = `number of chunks split by chunker` including duplicates
cardinality of `stored` = `number of distinct chunks stored`
therefore in most cases `stored <= total`.
I am still voting to get rid of these different cardinalities for the sake of other people in the team that could understand this.

Reply:
`STORED` is incremented on every call to put, `SEEN` incremented when `has==true` meaning: 
- `STORED` is the number of chunk instances (split by chunker) for which `Put` has returned
- `SEEN` is the number of chunk instances for which `Put` has returned and turned out to be a duplicate

it eludes me what is so difficult here that you feel like protecting the team from the damage caused.... 

it is total chunks vs distinct chunks, 
- `SPLIT` and `STORED` compare to `TOTAL` 
- `SENT` and `SYNCED` compare to total distinct = `TOTAL-SEEN` 

plus we indicate if these numbers are not yet accurate.
We should add a unit test to this function that simply checks this logic

Reply:
> it eludes me what is so difficult here that you feel like protecting the team from the damage caused....

evidently, you and I interpret `STORED` in 2 different ways. For me it is the actual number of chunks stored on persistent storage (`TOTAL-SEEN`). In any case, i'm mending this as per your requirements as I'd rather focus on other things right now :+1:

Reply:
changed this and added a unit test